google/cloud/internal/timer_queue.cc

```diff
@@ -46,80 +56,83 @@ void TimerQueue::Shutdown() {
   std::unique_lock<std::mutex> const lk(mu_);
   shutdown_ = true;
   cv_.notify_all();
+  cv_follower_.notify_all();
 }
 
-// Timers added via Schedule should be managed by one or more threads that
-// call Service. Calls to Service only return after Shutdown has been called.
 void TimerQueue::Service() {
-  while (true) {
-    std::unique_lock<std::mutex> lk(mu_);
-    auto const until = NextExpiration(lk);
-    // TODO(#10521): Predicate condition ne < until may need to be reworked.
-    auto predicate = [this, until, &lk] {
-      auto const ne = NextExpiration(lk);
-      return shutdown_ || ne < until ||
-             HasExpiredTimer(lk, std::chrono::system_clock::now());
+  // It is easier to understand this if you first ignore shutdown and empty
+  // timer queues.
+  //
+  // The threads play two different roles:
+  // - A single thread at a time plays the leader role.
+  // - All other threads are followers and block on cv_follower_.
+  //
+  // Once a timer expires the leader thread relinquishes its role and wakes up
+  // one follower thread to become the new leader. Only after doing so it
+  // runs the code to expire the timer.
+  //
+  // This is all complicated by shutdown. Basically all thread needs to wake
+  // up when the TQ is shutdown and one of them will expire all the timers.
+  //
+  std::unique_lock<std::mutex> lk(mu_);
+  auto is_leader = false;
+  while (!shutdown_) {
+    if (!is_leader && has_leader_) {
+      // The current thread becomes a follower until no leader is available.
```

Comment:
s/until no leader is available/while there is a leader/ perhaps?

Reply:
Done.